### PR TITLE
Add ShellCheck to OpenCorePkg CI

### PR DESCRIPTION
**Pull request by dakanji**
Closes acidanthera/bugtracker/#902

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -7,6 +7,18 @@ env:
...
   include:
+  - os: osx
+    name: "Check Shell Scripts"
+    osx_image: xcode10.2
+    compiler: clang
+
+    script:
+      - HOMEBREW_NO_AUTO_UPDATE=1 brew install shellcheck
+      # Check Parent Dir (No *.sh files curently. Change "*.tool" to "*.{sh,tool}" if any is added later)
+      - shellcheck /home/travis/build/acidanthera/OpenCorePkg/*.tool
```
Comment (vit9696) [inline .travis.yml:18]: Use `$(pwd)` if you want full path. We should not rely on travis internal paths.

File: .travis.yml [hunk 2]
```diff
@@ -7,6 +7,18 @@ env:
...
+      # Check Sub Dirs (No *.sh files curently. Change "*.tool" to "*.{sh,tool}" if any is added later)
+      - shellcheck /home/travis/build/acidanthera/OpenCorePkg/**/*.tool
```
Comment (vit9696) [inline .travis.yml:20]: Here too.

File: .travis.yml [hunk 3]
```diff
@@ -15,9 +15,9 @@ matrix:
     script:
       - HOMEBREW_NO_AUTO_UPDATE=1 brew install shellcheck
       # Check Parent Dir (No *.sh files curently. Change "*.tool" to "*.{sh,tool}" if any is added later)
-      - shellcheck /home/travis/build/acidanthera/OpenCorePkg/*.tool
+      - shellcheck ./*.tool
```
Comment (vit9696) [inline .travis.yml:18]: In fact, I think you want `shellcheck ./*.tool || exit 1` here, because otherwise the error may not be reported?

Comment (dakanji) [reply .travis.yml:18]: It will exit with "1" once there is an error.

**Comment by dakanji** [comment]
Can you advise whether `compiler: clang` and `osx_image: xcode10.2` are actually needed for this section?

**Comment by vit9696** [comment]
You could use Linux VMs really, as they are faster. E.g.

```
  - os: linux
    name: "Shellcheck on Linux"
    addons:
      apt:
        packages:
        - shellcheck
```

File: .travis.yml (hunk 3 above)
Comment (vit9696) [reply .travis.yml:19]: shellcheck will, but travis does not abort build on a command returning non-zero code.

Comment (dakanji) [reply .travis.yml:19]: It will continue the build but report that test as failed.
That way can see that Test A failed and Test D failed for amendments.

Comment (dakanji) [reply .travis.yml:19]: Please confirm.

Do you prefer to abort on the first error met (Your Suggested Change ... Should really then be done for each test. I.e., running macbuild etc should also have similar ) or continue and say so many tests failed (This is how everything is currently setup)?

Comment (vit9696) [reply .travis.yml:19]: I guess either way is fine to me in this case, I will merge.

**Comment by dakanji** [comment]
Found out that there are some shell scripts with a `command` extension that were not checked before and `/Utilities/LogoutHook/LogoutHook.command` has quite a few issues flagged starting with the shebang which is plain sh but there is one non POSIX compliant item in the code.

That one is easy to fix by changing to bash but will be adding exceptions for the others with links to the ShellCheck wiki as there is no test on this script and can't be sure of implications of changes.

Bearing your reluctance to open many issues, I think an issue on this should be opened for you to evaluate and remove the exceptions later.

Thoughts?

**Comment by vit9696** [comment]
Basically it is like this:

- We will add more scripts soon, e.g. I will likely move `https://github.com/acidanthera/Lilu/blob/master/Lilu/Scripts/kpdescribe.sh` to OpenCorePkg.
- I am ready to review changes and/or ask people to ensure that the scripts continue working after the changes.
- I am not ready to invest time into into shellcheck at least in the nearest future, so if it causes problems I will rather disable it entirely.
- We are slowly changing `bash` to `sh`, because newer macOS use `zsh` by default and we cannot rely on `bash` features anymore.

So if you could fix the script and submit a PR about it I can always help, but the rest is beyond our scope as we have important things to handle.

**Comment by dakanji** [comment]
OK. Will leave the exceptions in place on this one and they will be skipped.
Can be looked at when you are doing your changes.

Note though that shellcheck will catch the sh incompatibilities such as seen here and will help your reviewers later

**Pull request merged**